core/src/test/scala/com/mesosphere/usi/core/integration/SchedulerFuzzingTest.scala

```diff
@@ -0,0 +1,137 @@
+    .build()
+
+  "random commands" in {
+    // Expected pod states
+    val expectedState = mutable.Map[PodId, Protos.TaskState]()
+
+    // Observed pod states as reported by USI
+    var observedState = concurrent.TrieMap.empty[PodId, Protos.TaskState]
+
+    // Commands
+    def genPodLaunches =
+      for {
+        //podId <- Arbitrary.arbitrary[String].suchThat(s => s.nonEmpty && s.matches("^[a-zA-Z0-9\\-\\.]+$"))
+        podId <- Arbitrary.arbitrary[Int]
+        cpu <- Gen.choose(0.1, 1.1)
+        mem <- Gen.choose(2, 256)
+      } yield
+        LaunchPod(
+          PodId(s"random-pod-$podId"),
+          SimpleRunTemplateFactory(
+            resourceRequirements = List(ScalarRequirement.cpus(cpu), ScalarRequirement.memory(mem)),
+            shellCommand = "sleep 3600",
+            "test")
+        )
+
+    // Gen.delay will evaluate expectedState each time.
+    def genPodKills = Gen.delay {
+      if (expectedState.nonEmpty) Gen.oneOf(expectedState.keys).map(KillPod)
+      else Arbitrary.arbitrary[Int].map(id => KillPod(PodId(s"unknown-pod-$id")))
+    }
+
+    def genCommands = Gen.oneOf(genPodLaunches, genPodKills)
+
+    // Setup USI
+    lazy val mesosClient: MesosClient = MesosClient(mesosClientSettings, frameworkInfo).runWith(Sink.head).futureValue
+    val schedulerSettings = SchedulerSettings
+      .load()
+      .withDebounceReviveInterval(DurationConverters.toJava(50.millis))
+    lazy val factory = SchedulerFactory(mesosClient, InMemoryPodRecordRepository(), schedulerSettings, DummyMetrics)
+    lazy val (_, schedulerFlow) = factory.newSchedulerFlow().futureValue
+    lazy val (input, output) = commandInputSource
+      .log("scheduler commands")
+      .via(schedulerFlow)
+      .log("scheduler events")
+      .toMat(Sink.foreach(updateObservedState))(Keep.both)
+      .withAttributes(Attributes
+        .logLevels(onElement = Logging.DebugLevel, onFinish = Logging.InfoLevel, onFailure = Logging.ErrorLevel))
+      .run
+
+    // Run commands
+    forAll(genCommands) { cmd =>
+      logger.debug(s"Running command. command=$cmd")
+
+      updateExpectedState(cmd)
+
+      // Apply command
+      input.offer(cmd).futureValue
+
+      // Assert
+      eventually {
+        logger.debug("Asserting observed and expected state.")
+        observedState.snapshot() should contain theSameElementsAs expectedState
+      }
+    }
+
+    // Stop
+    mesosClient.killSwitch.shutdown()
+    output.futureValue
+
+    /**
+      * Updated the expected state. It is called in the beginning of forAll.
+      * @param cmd The next command.
+      */
+    def updateExpectedState(cmd: SchedulerCommand): Unit = cmd match {
+      case launchPod: LaunchPod => expectedState.update(launchPod.podId, Protos.TaskState.TASK_RUNNING)
+      case killPod: KillPod =>
+        // Killed pods are automatically expunged once they become terminal.
```

Comment:
I'm not sure this is expected behavior. I would expect that I have to expunge a pod when it is terminal.